dspinner/pin_test.go

```diff
@@ -0,0 +1,1056 @@
+
+func TestPinRecursiveFail(t *testing.T) {
+	ctx := context.Background()
+	dstore := dssync.MutexWrap(ds.NewMapDatastore())
+	bstore := blockstore.NewBlockstore(dstore)
+	bserv := bs.New(bstore, offline.Exchange(bstore))
+	dserv := mdag.NewDAGService(bserv)
+
+	p := New(dstore, dserv)
+
+	a, _ := randNode()
+	b, _ := randNode()
+	err := a.AddNodeLink("child", b)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// NOTE: This isnt a time based test, we expect the pin to fail
+	mctx, cancel := context.WithTimeout(ctx, time.Millisecond)
+	defer cancel()
+
+	err = p.Pin(mctx, a, true)
+	if err == nil {
+		t.Fatal("should have failed to pin here")
+	}
+
+	err = dserv.Add(ctx, b)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = dserv.Add(ctx, a)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// this one is time based... but shouldnt cause any issues
+	mctx, cancel = context.WithTimeout(ctx, time.Second)
+	defer cancel()
+	err = p.Pin(mctx, a, true)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestPinUpdate(t *testing.T) {
+	ctx := context.Background()
+
+	dstore := dssync.MutexWrap(ds.NewMapDatastore())
+	bstore := blockstore.NewBlockstore(dstore)
+	bserv := bs.New(bstore, offline.Exchange(bstore))
+
+	dserv := mdag.NewDAGService(bserv)
+	p := New(dstore, dserv)
+	n1, c1 := randNode()
+	n2, c2 := randNode()
+	_, c3 := randNode()
+
+	err := dserv.Add(ctx, n1)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err = dserv.Add(ctx, n2); err != nil {
+		t.Fatal(err)
+	}
+
+	if err = p.Pin(ctx, n1, true); err != nil {
+		t.Fatal(err)
+	}
+
+	if err = p.Update(ctx, c1, c2, true); err != nil {
+		t.Fatal(err)
+	}
+
+	assertPinned(t, p, c2, "c2 should be pinned now")
+	assertUnpinned(t, p, c1, "c1 should no longer be pinned")
+
+	if err = p.Update(ctx, c2, c1, false); err != nil {
+		t.Fatal(err)
+	}
+
+	// Test updating same pin that is already pinned.
+	if err = p.Update(ctx, c2, c2, true); err != nil {
+		t.Fatal(err)
+	}
+	// Check that pin is still pinned.
+	_, ok, err := p.IsPinned(ctx, c2)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !ok {
+		t.Fatal("c2 should still be pinned")
+	}
+
+	// Test updating same pin that is not pinned.
+	if err = p.Update(ctx, c3, c3, false); err == nil {
+		t.Fatal("expected error updating unpinned cid")
+	}
+	_, ok, err = p.IsPinned(ctx, c3)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if ok {
+		t.Fatal("c3 should not be pinned")
+	}
+
+	assertPinned(t, p, c2, "c2 should be pinned still")
+	assertPinned(t, p, c1, "c1 should be pinned now")
+}
+
+func TestLoadDirty(t *testing.T) {
+	ctx := context.Background()
+
+	dstore := dssync.MutexWrap(ds.NewMapDatastore())
+	bstore := blockstore.NewBlockstore(dstore)
+	bserv := bs.New(bstore, offline.Exchange(bstore))
+	dserv := mdag.NewDAGService(bserv)
+
+	p := New(dstore, dserv)
+
+	a, ak := randNode()
+	err := dserv.Add(ctx, a)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Pin A{} recursive
+	err = p.Pin(ctx, a, true)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	cidStr := ak.String()
+
+	// Corrupt index
```

Comment:
We should corrupt the index with an insertion as well as a deletion

Reply:
Done.